### PR TITLE
Adding in Code for Configuration/Settings Handling

### LINKED ISSUE #24: Retrieve Client Configuration for ManageIQ Authentication
Provide a way for the client to retrieve client configuration defined by the user.
(Need ManageIQ host and credentials -> token or username/password, enabling ssl, etc...)

### PR DESCRIPTION
**Pull request by vi-patel**
Precedence of Gathering the Configuration:
1. passing global options
2. ENV_VAR that is a dictionary of the required keys (MIQ_CFG)
3. local file
4. MIQCLI_CFGFILE (/etc/miqcli/miqcli.cfg)

This pull request closes #24

### CODE REVIEW

File: requirements.txt
Comment (ryankwilliams) [inline requirements.txt]: Could we state the version of PyYAML to be installed? I.e:

`PyYAML==3.12`

File: example/miqcli.cfg
Comment (tehsmyers) [inline example/miqcli.cfg]: Any reason not to name this miqcli.yml? Maybe this is just me, but I think `.yml`/`.yaml` means "config file", and `.cfg` means "probably like an ini file, or otherwise some structure unique to whatever it is that I'm configuring".

Throwing a yaml extension on this would help make it intuitively clear to users (and the editors they use) that this is a yaml file.

File: miqcli/cli/collection.py
Comment (tehsmyers) [inline miqcli/cli/collection.py]: Need two empty lines before a class def

File: miqcli/cli/collection.py
Comment (ryankwilliams) [inline miqcli/cli/collection.py]: When setting the exit code, could we use the following:

`raise SystemExit(1)`

sys.exit(arg) raises SystemExit exception. This would remove the need to include an additional library.

Comment (tehsmyers) [inline miqcli/cli/collection.py]: All prints in the CLI should probably be done with [`click.echo`](http://click.pocoo.org/5/quickstart/#echoing), and all fails should probably be done with click [exception handling](http://click.pocoo.org/5/exceptions/#exception-handling) to ensure that errors end up going out via `sys.stderr` and exiting the program in a consistent way.

File: miqcli/cli/collection.py
Comment (tehsmyers) [inline miqcli/cli/collection.py]: This is a "diaper" antipattern; please be explicit about what exceptions this is prepared to handle.

File: miqcli/cli/collection.py
Comment (ryankwilliams) [inline miqcli/cli/collection.py]: Maybe a local variable could be defined setting up the absolute path for the configuration file. Then both open statements could reference the variable instead of creating the absolute path to the file. Just a suggestion. Removes the need if the constant names ever changed down the road. Modifications to one place instead of two.

File: miqcli/cli/collection.py
Comment (ryankwilliams) [inline miqcli/cli/collection.py]: If we decide to use open to read files, we should close the file once finished. A suggestion could be to use with keyword to handle both of these:

```python
with open(local_cfg, 'r') as f:
    self._settings = yaml.load(f)
```
This could be applied to all open statements in the module.

File: miqcli/cli/collection.py
Comment (tehsmyers) [inline miqcli/cli/collection.py]: It's unclear to me how settings are intended to be set by this method, since it never actually "sets" them, it only validates them based on what was passed in the contructor. Based on the docstring, this should probably be named "validate", or something similar.

If we want a `set_` method to do this validation, it should probably just be a setter method, e.g.
```python
@settings.setter
def settings(value):
```
This means that the init line would likely become `self.settings = settings`, with the benefit that `Collection` users don't have to remember to call `set_settings` to ensure their settings are validated.

File: miqcli/cli/collection.py
Comment (ryankwilliams) [inline miqcli/cli/collection.py]: Could we switch to using raise SystemExit(1). sys.exit(arg) actually raises SystemExit. It just removes the need of an additional library. Just a suggestion.

```python
raise SystemExit(1)
```

**Review by ryankwilliams — CHANGES_REQUESTED** [review]
Overall looks good to me for being able to pass the server authentication details to the collections. Each collection will require this data when it goes to establish a connection to ManageIQ. I can see this working with #23 when needing authentication details.

I added two suggestions and one item that would be nice to change.

**Review by ryankwilliams — APPROVED** [review]

File: miqcli/collections/actions.py
Comment (tehsmyers) [inline miqcli/collections/actions.py]: In order to maintain a clear separation of concerns between the "library" components and the CLI, at no point should any module in the `miqcli.collections` namespace depend on anything from the `miqcli.cli` namespace. I think it's a great idea to have a `Collection` base class, but that class should probably be defined in `miqcli.collections.base` or some namespace along those lines.

**Review by tehsmyers — CHANGES_REQUESTED** [review]
I think it should be possible to bake the config file handling directly into the `ManageIQ` `click.MultiCommand` subclass. It makes sense that the REST interface classes should be able to validate the settings they're given, but on the CLI side it would be good to centralize all settings inputs through the `ManageIQ` class.

File: miqcli/collections/actions.py
Comment (ryankwilliams) [reply miqcli/collections/actions.py]: The more I think about this, I agree with @seandst on this suggestion. I was back and forth on the location with the class also. Having this separation will maintain a clean code base structure.

**Comment by ryankwilliams** [comment]
Just a thought:

All of the collections modules will need the authentication information that the ManageIQ CLI class collects. There could be an additional method within that class to collect this information in the ways that it can be defined. This method could return the authentication information that is then passed to the Collections class when it is instantiated. It would be okay to pass null values for authentication to the collection class because the collection would be the one to validate it when it goes to establish a connection to ManageIQ server. The collection would then have the responsibility to say "Authentication failures, please check setting your credentials" when connecting.

File: miqcli/collections/actions.py
Comment (vi-patel) [reply miqcli/collections/actions.py]: Thanks for the reviews, all the comments made sense, I have made the suggested updates.

**Review by tehsmyers — APPROVED** [review]

**Comment by tehsmyers** [comment]
Please remember to squash before merging, and thanks!

Also, I'm not sure if we're going with a "last reviewer merges" approach or "PR author merges if possible". Let me know if I'm the one that should hit the big green button.

**Comment by tvieira** [comment]
Sorry, I think I'm too late to comment in this PR. But even though I would like to comment on the order of precedence. Configuration files within etc are usually default configuration or, if added by the user after system is installed, it is the first resource to be looked at. Then, any configuration added to the user local environment supersede the global settings. And any other settings added in the command line call supersede the the global and user local settings. Otherwise, if you run the client into a system that is not administrated by the user, the user won't be able to supersede configuration added by the admin. I don't think this would be a desirable output. Does it make sense? Let me know wyt.

**Comment by tehsmyers** [comment]
That was my understanding of [this docstring](https://github.com/rhpit/manageiq-cli/pull/28/files#diff-bb83cfdafdba9cdcb2267e847eb91292R135): command-line options have the final say, followed by the `MIQ_CFG` envvar, and then defaults come from the config files (first local, then systemwide).

I'm not sure that the code in this PR does exactly that (I haven't taken a close enough look), but I think the intent is clear. We can refine the implementation in this PR or following PRs as needed.

Sorta related, by click has built-in support for config via envvar that we might also consider implementing, though having this mechanism in place alongside the `MIQ_CFG` may just end up blinding users with too many options.

**Review by tvieira — CHANGES_REQUESTED** [review]
I would prefer to have overwriting by precedence, specific parameters passed via command line overwrites the local parameters which overwrite the global parameters.

File: miqcli/cli/main.py
Comment (tvieira) [inline miqcli/cli/main.py]: help doesn't match option

File: miqcli/cli/main.py
Comment (tvieira) [inline miqcli/cli/main.py]: help doesn't match option

**Comment by tvieira** [comment]
based on this documentation from ManageIQ
http://manageiq.org/docs/reference/fine/api/overview/auth

> `Expire after a certain amount of time with no REST API activity - 10 minutes by default.`

With that in mind, it doesn't make sense to set that in the configuration file. We should have a token set in a temporary file to assist the command line to connect to the ManageIQ instance. But this is the responsibility of the CLI and not the user. The user has to provide only the username, password and the host wheren ManageIQ is installed.

The order for the token should be something like:
```
if a $HOME/.miqcli/auth file exist
  validate token
  if token doesn't work
    create new token
    update $HOME/.miqcli/auth
else
  create  $HOME/.miqcli/auth
  create new token
  update $HOME/.miqcli/auth
use token from $HOME/.miqcli/auth
```

File: miqcli/cli/main.py
Comment (tvieira) [inline miqcli/cli/main.py]: you don't need the check_yaml file anymore. from_yaml function does the work for you. If you set silent=True nothing happens if the file doesn't exist or if the file is invalid. Maybe silent should be set to false? Either way, you don't need check_yaml anymore. That can be removed from the source code.

File: miqcli/cli/main.py
Comment (tvieira) [inline miqcli/cli/main.py]: Same case of the item above.

File: miqcli/cli/main.py
Comment (tvieira) [inline miqcli/cli/main.py]: same case. The function from_env will check automatically. If the environment variable is not set, nothing happens. If the environment variable is set, it loads the content. If silent is set to True and the content is wrong, nothing happens. If the silent is set to False and the content is wrong, then a failure will be raised.

**Review by tvieira — CHANGES_REQUESTED** [review]
I don't see the default settings for username, password and url. Why we don't leave the settings if we are running with the docker images? (username=admin, password=smartvm, url=lhttps://ocalhost:8443)?

File: miqcli/cli/main.py
Comment (tvieira) [inline miqcli/cli/main.py]: again, we don't need check_yaml function.

File: miqcli/cli/main.py
Comment (tvieira) [inline miqcli/cli/main.py]: same as above.

**Review by tvieira — CHANGES_REQUESTED** [review]
I would remove check_yaml, it does not add any value.

**Comment by vi-patel** [comment]
@tvieira, check_yaml helps me determine the config file name regardless of the extension .yml or .yaml.  Do you want me to move that code into the from_yaml method of the Config class?

**Comment by tvieira** [comment]
@vi-patel I think this is too much just to deal with extension. Why we don't use a constant that has the value `/etc/miqcli/config` and we keep it simple? What do you think?

**Review by tvieira — APPROVED** [review]
LGTM

**Pull request merged**
